### DISCUSSION TITLE
Yet another "No patterns detected"

### DISCUSSION THREAD

**Discussion by keltik85** [BennyThadikaran/stock-pattern#91] [Q&A]
The command i start the script with: `python init.py --tf 60 -p all --save -r 1 -l 1`

The config i use with the script: 
```
{
  "DATA_PATH": "binance_klines",
  "LOADER": "IEODFileLoader",
  "DEFAULT_TF": "1",
  "DATE_COLUMN": "timestamp",
  "SYM_LIST": "symbollist.txt",
  "24_7": true
}
```

The content of `symbollist.txt`:
```
BONKUSDT
```

I dont paste the content of `src/binance_klines/bonkusdt.csv`, because it contains 1000 lines. I added the file to this discussion via the upload:
[bonkusdt.csv](https://github.com/user-attachments/files/17919227/bonkusdt.csv)

With these inputs the result is: 
```
Scanning `ALL` patterns on `60`. Press Ctrl - C to exit
100%|██████████| 1/1 [00:01<00:00,  1.92s/it]
No patterns detected
```

I read this discussion already: https://github.com/BennyThadikaran/stock-pattern/discussions/69
and added the `-r 1` and `-l 1` flags to my command, but still there is nothing detected.

Why is there no pattern detected at all?

**Comment by BennyThadikaran** [comment]
You have set your `DEFAULT_TF` to `1`. It means 1 min data, whereas your data is 60min.

So here is the correct config.
```json
{
  "DATE_COLUMN": "timestamp",
  "DATA_PATH": "/home/benny/Desktop/data",
  "LOADER": "IEODFileLoader",
  "DEFAULT_TF": "60",
  "24_7": true
}
```

I tried plotting the data and found a head & shoulders in the past as marked in the image below.

![Figure_1](https://github.com/user-attachments/assets/465264a4-7896-48ad-bc32-9c4af4f05936)

1. Only the last 160 candles are loaded for scanning. The pattern must end near the last candle or atleast must be valid near the last candle.

2. Accordingly, I provide a data close to the pattern end to detect the pattern.

Below I provide the time portion. If only `2024-11-23` is provided it is interpreted as 2024-11-23T00:00. No data after this point is loaded.

**So try running the below**

`py init.py -p all --sym bonkusdt -d "2024-11-23T11:00"`

Finally the last pivot point must be 6 candles unbroken to be confirmed as a pivot. That is what the `-r` and `-l` option does.  

Using just 1 candle pivots will create a lots of pivots that have no significance in terms of support or resistance. 

Hope this helps.

**Comment by keltik85** [comment]
> py init.py -p all --sym bonkusdt -d "2024-11-23T11:00"

When i run it i get this output:

![image](https://github.com/user-attachments/assets/e2fe792b-4c22-4244-ac73-48b447db76c4)

So there is no head and shoulders. There is a VCP detected.

When i change the date to this `py init.py -p all --sym bonkusdt -d "2024-11-19T19:00"`

I again get `No patterns detected`, which strikes me as odd, because you wrote that:
> Only the last 160 candles are loaded for scanning. 

The date `2024-11-19T19:00` is the 159th counting from the bottom of the csv.


### UPDATE - 01:
When i run the command with this date: `python init.py --tf 60 -p all --save -d 2024-11-24T07:00`.

I again get the above VCP in the picture.

### UPDATE - 02
However i have to add that i start the script with `python 3.8` and had to reimplement itertools inside the Plotter class:

```
    @staticmethod
    def pairwise(iterable):
        """s -> (s0, s1), (s1, s2), (s2, s3), ..."""
        a, b = itertools.tee(iterable)
        next(b, None)
        return zip(a, b)
```

the itertools.pairwaise function is only available after python 3.10.

### UPDATE - 03:
After trying it out with python 3.10 i still get the VCP picture.

**Reply by BennyThadikaran** [reply]
Thanks for point out the pairwise issue. I will double check the python 3.8 support. 

My bad :stuck_out_tongue: , it was indeed a VCP that was detected on mine too. I was just confirming my own bias after checking the chart visually.

As for the VCP, it has nothing to do with the issues you mentioned. We are missing the left head of the HNS. 

If you look at the image you posted, 

- the first candle could serve as a pivot for the left head, but there no prior candles to confirm the pivot point. (You need 6 candles prior and 6 after to confirm the pivot)
- the seventh candles (red pin bar) could serve as a pivot but it has only 4 prior candles that have not violated the low.
- I did try matching with `-l 4 -r 4` but it was not detected. I suppose the pivots are too far apart or not aligned enough to be detected. (I need to look into this further.)

Since there are no pivots that match the HNS, it was not detected and the next matching pattern was the VCP.

So like i mentioned in the README, this cannot replace a traders trained eye but is just an assistance tool. It wont detect every single pattern but only ones that fall within the framework of rules and `filters.

Anyways this was a fruitful discussion. should serve as guide for others to understand how the program works.

**Reply by BennyThadikaran** [reply]
Version 4.0.5 resolves the issue with itertools.pairwise in Python 3.8 and 3.9. I tested it on Python 3.8.10. All working well.
